src/main/scala/org/hammerlab/guacamole/alignment/AffineGapPenaltyAlignment.scala

```diff
@@ -0,0 +1,159 @@
+                          mismatchProbability: Double,
+                          openGapProbability: Double,
+                          closeGapProbability: Double): (DenseMatrix[AlignmentState], DenseMatrix[Double]) = {
+
+    val logMismatchPenalty = -math.log(mismatchProbability)
+    val logOpenGapPenalty = -math.log(openGapProbability)
+    val logCloseGapPenalty = -math.log(closeGapProbability)
+
+    val logContinueGapPenalty = -math.log(1 - closeGapProbability)
+    val noGapPenalty = -math.log(1 - openGapProbability)
+
+    val sequenceLength = sequence.length
+    val referenceLength = reference.length
+
+    // alignmentScores is M x N matrix
+    // M is the length of the sequence and N is the length of the reference
+    val alignmentScores = DenseMatrix.fill[Double](sequenceLength + 1, referenceLength + 1) { Double.PositiveInfinity }
+    val alignmentStates = new DenseMatrix[AlignmentState](sequenceLength + 1, referenceLength + 1)
+
+    def transitionPenalty(nextState: AlignmentState, previousState: AlignmentState, isEndState: Boolean) = {
+
+      val openGap = nextState != previousState && AlignmentState.isGapAlignment(nextState)
+      val closeGap = nextState != previousState && AlignmentState.isGapAlignment(previousState)
+      val continueGap = nextState == previousState && AlignmentState.isGapAlignment(nextState)
+      val mismatch = nextState == AlignmentState.Mismatch
+
+      (if (openGap) logOpenGapPenalty else 0) +
+        (if (closeGap) logCloseGapPenalty else 0) +
+        (if (continueGap) logContinueGapPenalty else if (mismatch) noGapPenalty + logMismatchPenalty else noGapPenalty) +
+        (if (isEndState && AlignmentState.isGapAlignment(nextState)) logCloseGapPenalty else 0)
+    }
+
+    // empty case
+    alignmentScores(0, 0) = 0
+
+    for (referenceIdx <- 1 to referenceLength) {
+      for (sequenceIdx <- 1 to sequenceLength) {
+        // Given the change in position, is the transition a gap or match/mismatch
+        def classifyTransition(prevSeqPos: Int, prevRefPos: Int): AlignmentState = {
+          if (sequenceIdx == prevSeqPos) {
+            AlignmentState.Deletion
+          } else if (referenceIdx == prevRefPos) {
+            AlignmentState.Insertion
+          } else if (sequence(sequenceIdx - 1) != reference(referenceIdx - 1)) {
+            AlignmentState.Mismatch
+          } else {
+            AlignmentState.Match
+          }
+        }
+
+        val possiblePreviousStates = Seq(
+          (sequenceIdx - 1, referenceIdx),
+          (sequenceIdx, referenceIdx - 1),
+          (sequenceIdx - 1, referenceIdx - 1)
+        )
+
+        // Compute the transition costs based on the gap penalties
+        val stateScores: Seq[(AlignmentState, Double)] = possiblePreviousStates.map({
+          case (prevSeqPos, prevRefPos) => {
+            val nextState = classifyTransition(prevSeqPos, prevRefPos)
+            val previousState = alignmentStates(prevSeqPos, prevRefPos)
```

Comment:
Yes, you are right here, it was relying on that behavior accidently, but you've fixed this in your PR